File: src/provide/telemetry/consent.py
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""Consent-aware telemetry collection."""

from __future__ import annotations

__all__ = [
    "ConsentLevel",
    "get_consent_level",
    "set_consent_level",
    "should_allow",
]

import enum
import os
import threading
import warnings

from provide.telemetry.levels import LogSeverity, level_order

_CONSENT_ENV_VAR = "PROVIDE_CONSENT_LEVEL"


class ConsentLevel(enum.Enum):
    FULL = "FULL"
    FUNCTIONAL = "FUNCTIONAL"
    MINIMAL = "MINIMAL"
    NONE = "NONE"


_lock = threading.Lock()
_level: ConsentLevel = ConsentLevel.FULL
_invalid_env_warned = False


def set_consent_level(level: ConsentLevel) -> None:
    global _level
    with _lock:
        _level = level


def get_consent_level() -> ConsentLevel:
    with _lock:
        return _level


def _rank(log_level: str | None) -> int:
    """Order a log level for consent comparisons.

    Resolves through the one shared table. An unrecognised level now ranks INFO
    rather than the old local default of 0/TRACE; both sit below the WARN and
    ERROR gates below, so no consent decision changes. FATAL does change: it
    used to be unrecognised and was dropped as if it were the least severe
    record in the ladder.

    The two hoisted pragma constants this replaced -- a placeholder for the
    missing-level lookup and its 0 rank -- went with the dict they served.
    """
    return level_order(log_level)


def should_allow(signal: str, log_level: str | None = None) -> bool:
    with _lock:
        level = _level

    if level == ConsentLevel.FULL:
        return True
    if level == ConsentLevel.NONE:
        return False
    if level == ConsentLevel.FUNCTIONAL:
        if signal == "logs":
            return _rank(log_level) >= LogSeverity.WARN
        return signal != "context"  # traces and metrics allowed; context blocked
    # MINIMAL
    if signal == "logs":
        return _rank(log_level) >= LogSeverity.ERROR
    return False  # traces/metrics/context blocked at MINIMAL


def _load_consent_from_env() -> None:
    """Apply ``PROVIDE_CONSENT_LEVEL`` if it is set.

    Called by ``setup_telemetry()`` and by the lazy ``get_logger()`` path, so an
    operator opt-out takes effect without a code change. Trimmed and upper-cased.
    An unset or blank variable is a no-op (a level chosen in code survives). A
    set, non-empty, unrecognised value fails closed: consent becomes ``NONE``
    and a ``RuntimeWarning`` naming the value is emitted once per process. The
    variable is an opt-out control, and the one failure an opt-out must not
    have is a typo that silently leaves collection on.
    """
    raw = os.environ.get(_CONSENT_ENV_VAR)
    if raw is None:
        return
    text = raw.strip()
    if not text:
        return
    try:
        level = ConsentLevel(text.upper())
    except ValueError:
        level = ConsentLevel.NONE
        _warn_invalid_consent_env_once(raw)
    set_consent_level(level)


def _warn_invalid_consent_env_once(raw: str) -> None:
    global _invalid_env_warned
    with _lock:
        if _invalid_env_warned:
            return
        _invalid_env_warned = True
    message = (
        f"{_CONSENT_ENV_VAR}={raw!r} is not one of FULL, FUNCTIONAL, MINIMAL, NONE; consent set to NONE (fail-closed)"
    )
    warnings.warn(message, RuntimeWarning, stacklevel=3)  # pragma: no mutate — stacklevel is cosmetic


def _reset_consent_for_tests() -> None:
    global _invalid_env_warned, _level
    with _lock:
        _level = ConsentLevel.FULL
        _invalid_env_warned = False  # pragma: no mutate — None is an equivalent falsy reset

File: typescript/src/consent.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * Consent-aware telemetry collection.
 * When deleted, all signals pass through unchanged.
 */

import { LogSeverity, levelOrder } from './levels.js';

export type ConsentLevel = 'FULL' | 'FUNCTIONAL' | 'MINIMAL' | 'NONE';

// Stryker disable next-line StringLiteral: initial value is overwritten by resetConsentForTests before any test observes it
let _consentLevel: ConsentLevel = 'FULL';

export function setConsentLevel(level: ConsentLevel): void {
  _consentLevel = level;
}

export function getConsentLevel(): ConsentLevel {
  return _consentLevel;
}

export function shouldAllow(signal: string, logLevel?: string): boolean {
  const level = _consentLevel;
  if (level === 'FULL') return true;
  if (level === 'NONE') return false;
  if (level === 'FUNCTIONAL') {
    if (signal === 'logs') {
      return levelOrder(logLevel) >= LogSeverity.Warn;
    }
    if (signal === 'context') return false;
    return true;
  }
  // MINIMAL
  if (signal === 'logs') {
    return levelOrder(logLevel) >= LogSeverity.Error;
  }
  return false;
}

const _CONSENT_ENV_VAR = 'PROVIDE_CONSENT_LEVEL';
const _VALID_LEVELS: readonly ConsentLevel[] = ['FULL', 'FUNCTIONAL', 'MINIMAL', 'NONE'];

/** Read PROVIDE_CONSENT_LEVEL, or undefined when unset or outside a Node-like runtime. */
// Stryker disable BlockStatement
function _rawConsentEnv(): string | undefined {
  try {
    // process.env is not available in browser builds after tree-shaking,
    // but some bundlers (esbuild, Vite) leave process.env.X inline replacements.
    // Stryker disable next-line ConditionalExpression,StringLiteral: process is always defined in Node.js/test environments
    return typeof process !== 'undefined' ? process.env[_CONSENT_ENV_VAR] : undefined;
  } catch {
    return undefined;
  }
}
// Stryker enable BlockStatement

/**
 * Apply `PROVIDE_CONSENT_LEVEL` to the active consent level.
 *
 * The value is trimmed and upper-cased; FULL, FUNCTIONAL, MINIMAL and NONE
 * are applied. An unset or blank (empty or whitespace-only) variable leaves
 * the current level untouched — it never resets to FULL — so a level set
 * programmatically survives a later call from setup or the lazy logger. A
 * set, non-empty, unrecognised value fails closed: consent becomes NONE and a
 * warning naming the raw value goes to `console.warn` once per process,
 * outside the SDK's own logger so the NONE it just applied cannot drop it.
 * The variable is an opt-out control, and the one failure an opt-out must not
 * have is a typo that silently leaves collection on. Every SDK shares this
 * rule.
 *
 * Called by setupTelemetry / setupTelemetryAsync, and by the lazy logger path
 * before any setup has run.
 */
export function loadConsentFromEnv(): void {
  const raw = _rawConsentEnv();
  if (raw === undefined) return;
  const candidate = raw.trim().toUpperCase();
  if (candidate === '') return;
  if ((_VALID_LEVELS as readonly string[]).includes(candidate)) {
    _consentLevel = candidate as ConsentLevel;
    return;
  }
  _consentLevel = 'NONE';
  _warnInvalidConsentEnvOnce(raw);
}

let _invalidEnvWarned = false;

function _warnInvalidConsentEnvOnce(raw: string): void {
  if (_invalidEnvWarned) return;
  _invalidEnvWarned = true;
  console.warn(
    `[provide-telemetry] ${_CONSENT_ENV_VAR}="${raw}" is not one of FULL, FUNCTIONAL, MINIMAL, NONE; consent set to NONE (fail-closed)`,
  );
}

export function resetConsentForTests(): void {
  _consentLevel = 'FULL';
  _invalidEnvWarned = false;
}
